src/Nethermind/Nethermind.Evm/Tracing/GasEstimator.cs

```diff
@@ -31,18 +29,20 @@ public class GasEstimator
             _blocksConfig = blocksConfig;
         }
 
-        public long Estimate(Transaction tx, BlockHeader header, EstimateGasTracer gasTracer)
+        public long Estimate(Transaction tx, BlockHeader header, EstimateGasTracer gasTracer, CancellationToken token = new())
         {
             IReleaseSpec releaseSpec = _specProvider.GetSpec(header.Number + 1, header.Timestamp + _blocksConfig.SecondsPerSlot);
 
-            long intrinsicGas = tx.GasLimit - gasTracer.IntrinsicGasAt;
-            if (tx.GasLimit > header.GasLimit)
+            tx.SenderAddress ??= Address.Zero; // If sender is not specified, use zero address.
+            tx.GasLimit = Math.Min(tx.GasLimit, header.GasLimit); // Limit Gas to the header
+
+            // Calculate and return additional gas required in case of insufficient funds.
+            UInt256 senderBalance = _stateProvider.GetBalance(tx.SenderAddress);
+            if (tx.Value != UInt256.Zero && tx.Value >= senderBalance)
```

Comment:
Should this include `IntrinsicGas` too?

Reply:
It wasn't part of the original code. I just moved the code around and made no changes to the calculation.